libnopegl/src/node_colorstats.c

```diff
@@ -0,0 +1,413 @@
+     * use Y=1 and Z=1. 128 remains an always safe value so we use that as
+     * a fallback.
+     */
+    const struct gpu_limits *limits = &gpu_ctx->limits;
+    const int max_group_size_x = limits->max_compute_work_group_size[0];
+    s->group_size = max_group_size_x >= 256 ? 256 : 128;
+    LOG(DEBUG, "using a workgroup size of %d", s->group_size);
+
+    s->pipeline_compat_init     = ngli_pipeline_compat_create(gpu_ctx);
+    s->pipeline_compat_waveform = ngli_pipeline_compat_create(gpu_ctx);
+    s->pipeline_compat_sumscale = ngli_pipeline_compat_create(gpu_ctx);
+    if (!s->pipeline_compat_init || !s->pipeline_compat_waveform || !s->pipeline_compat_sumscale)
+        return NGL_ERROR_MEMORY;
+
+    s->crafter_init     = ngli_pgcraft_create(ctx);
+    s->crafter_waveform = ngli_pgcraft_create(ctx);
+    s->crafter_sumscale = ngli_pgcraft_create(ctx);
+    if (!s->crafter_init || !s->crafter_waveform || !s->crafter_sumscale)
+        return NGL_ERROR_MEMORY;
+
+    const struct pgcraft_block block = {
+        .name     = "stats",
+        .type     = NGLI_TYPE_STORAGE_BUFFER,
+        .stage    = NGLI_PROGRAM_SHADER_COMP,
+        .writable = 1,
+        .block    = &s->blk.block,
+        .buffer   = s->blk.buffer,
+    };
+
+    int ret;
+    if ((ret = setup_init_compute(s, &block)) < 0 ||
+        (ret = setup_waveform_compute(s, &block, o->texture_node)) < 0 ||
+        (ret = setup_sumscale_compute(s, &block)) < 0)
+        return ret;
+
+    return 0;
+}
+
+static int init_block(struct colorstats_priv *s, struct gpu_ctx *gpu_ctx)
+{
+    s->blk.data = NULL; /* We do not have any CPU data */
+
+    struct block *block = &s->blk.block;
+    ngli_block_init(block, NGLI_BLOCK_LAYOUT_STD430);
+
+    static const struct {
+        const char *name;
+        int type;
+        int count;
+    } block_layout[] = {
+        {"max_rgb",       NGLI_TYPE_UVEC2, 0},
+        {"max_luma",      NGLI_TYPE_UVEC2, 0},
+        {"depth",         NGLI_TYPE_I32,   0},
+        {"length_minus1", NGLI_TYPE_I32,   0},
+        {"summary",       NGLI_TYPE_UVEC4, 1 << MAX_BIT_DEPTH},
+        /*
+         * The size of the block will depend on the texture size, which will
+         * only be known in the update callback for textures fed by a Media
+         * node. We use a variadic buffer because don't want to wait for the
+         * texture update to compile the shaders and prepare the pipelines.
+         */
+        {"data",          NGLI_TYPE_UVEC4, NGLI_BLOCK_VARIADIC_COUNT},
+    };
+    for (int i = 0; i < NGLI_ARRAY_NB(block_layout); i++) {
+        int ret = ngli_block_add_field(block, block_layout[i].name, block_layout[i].type, block_layout[i].count);
+        if (ret < 0)
+            return ret;
+    }
+
+    s->blk.buffer = ngli_buffer_create(gpu_ctx);
+    if (!s->blk.buffer)
+        return NGL_ERROR_MEMORY;
+
+    s->blk.usage |= NGLI_BUFFER_USAGE_STORAGE_BUFFER_BIT;
+    s->blk.buffer->usage |= NGLI_BUFFER_USAGE_STORAGE_BUFFER_BIT;
```

Comment:
Comment added:
```c
    /* Colorstats needs to write into the block so we bind it as SSBO */
    s->blk.usage = NGLI_BUFFER_USAGE_STORAGE_BUFFER_BIT;

    s->blk.buffer = ngli_buffer_create(gpu_ctx);
    if (!s->blk.buffer)
        return NGL_ERROR_MEMORY;

    /*
     * The size of the buffer depends on the texture size, which will only be
     * known in the update callback for textures fed by a Media node. We use a
     * variadic buffer size because don't want to wait for the texture update to
     * compile the shaders and prepare the pipelines.
     *
     * The init and prepare callbacks are still too early for this buffer
     * allocation, but we still need to set its usage for the pipeline bindings.
     */
    s->blk.buffer->usage |= NGLI_BUFFER_USAGE_STORAGE_BUFFER_BIT;
```